src/libsyntax/ext/deriving/generic/mod.rs

```diff
@@ -706,13 +710,30 @@ impl<'a> TraitDef<'a> {
     }
 }
 
-fn variant_to_pat(cx: &mut ExtCtxt, sp: Span, enum_ident: ast::Ident, variant: &ast::Variant)
-                  -> P<ast::Pat> {
-    let path = cx.path(sp, vec![enum_ident, variant.node.name]);
-    cx.pat(sp, match variant.node.kind {
-        ast::TupleVariantKind(..) => ast::PatEnum(path, None),
-        ast::StructVariantKind(..) => ast::PatStruct(path, Vec::new(), true),
-    })
+fn find_repr_type_name(diagnostic: &SpanHandler,
```

Comment:
The most conservative solution would be to change the intrinsic to return `(sign, abs-value)` to avoid having to figure out the repr at all, but that does make the comparisons a bit less efficient depending on how smart LLVM can be.